sql: TestRaceWithBackfill failed

sql.TestRaceWithBackfill [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16825247?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16825247?buildTab=artifacts#/) on release-23.2 @ [b40936bb179fdfaa133f22b983905ff021ff980c](https://github.com/cockroachdb/cockroach/commits/b40936bb179fdfaa133f22b983905ff021ff980c):


```
=== RUN   TestRaceWithBackfill
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/17b9b212a9f8f930df1697330cbbbb63/logTestRaceWithBackfill716899568
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/sql_test/pkg/sql/schema_changer_test.go:613: (TestRaceWithBackfill)
	NOTICE: .DB() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().DB() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/testutils/testcluster/testcluster.go:1881: ((*TestCluster).SplitTable)
	NOTICE: .SplitRange() called via implicit interface StorageLayerInterface;
HINT: consider using .StorageLayer().SplitRange() instead.
    schema_changer_test.go:341: schema change ALTER TABLE t.test ADD COLUMN x DECIMAL DEFAULT (DECIMAL '1.4') CHECK (x >= 0) took 800.432074ms
    schema_changer_test.go:341: schema change ALTER TABLE t.test DROP pi took 2.070169248s
    schema_changer_test.go:341: schema change CREATE UNIQUE INDEX foo ON t.test (v) took 636.486222ms
    schema_changer_test.go:339: pq: duplicate key value violates unique constraint "bar"
    schema_changer_test.go:341: schema change CREATE INDEX bar ON t.test(k) STORING (v) took 610.275714ms
    schema_changer_test.go:417: condition failed to evaluate within 45s: from sql_test_utils.go:112: expected 16044 key value pairs, but got 12033
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/17b9b212a9f8f930df1697330cbbbb63/logTestRaceWithBackfill716899568
--- FAIL: TestRaceWithBackfill (54.43s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaceWithBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-42052

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -646,14 +646,17 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 
 	ctx := context.Background()
 
-	// number of keys == 2 * number of rows; 1 column family and 1 index entry
-	// for each row.
-	if err := sqltestutils.CheckTableKeyCount(ctx, kvDB, codec, 2, maxValue); err != nil {
-		t.Fatal(err)
-	}
-	if err := sqlutils.RunScrub(sqlDB, "t", "test"); err != nil {
-		t.Fatal(err)
-	}
+	testutils.SucceedsSoon(t, func() error {
+		// number of keys == 2 * number of rows; 1 column family and 1 index entry
+		// for each row.
+		if err := sqltestutils.CheckTableKeyCount(ctx, kvDB, codec, 2, maxValue); err != nil {
+			return err
+		}
+		if err := sqlutils.RunScrub(sqlDB, "t", "test"); err != nil {
+			return err
+		}
+		return nil
+	})
 
 	// Run some schema changes with operations.
 

